### PR TITLE
Redmine#2504: merge ifvarclass/if FnCall and scalar correctly

### PR DESCRIPTION
**Pull request by tzz**
@nickanderson @kacfengine 

This fixes bug exposed by https://github.com/cfengine/masterfiles/pull/503#issuecomment-118873667

With acceptance test.

### CODE REVIEW

**Comment by nickanderson** [comment]
I can confirm that with your code the test you supplied passes and without it it fails.

**Comment by jimis** [comment]
I had a look, but I'm not comfortable with the evaluator, especially function calling. I'll leave this for @kacfengine.

@tzz what does it fix? A ProgrammingError in case we use function calls and same class as the class guard?

**Comment by tzz** [comment]
Yup @jimis it fixes the case where the `ifvarclass` is a FnCall, merging it with the desired one.  The old code only handled the case where it was a string.

The synthetically constructed FnCall in the merge case is `and("stringcontext", old-fn-call(...))`.

The code is not generic, it's specifically built to resolve this case.

**Comment by jimis** [comment]
Thanks @tzz. A comment in the beginning of the acceptance test would be
great! :-)

**Comment by tzz** [comment]
OK; added.

File: libpromises/policy.c [hunk 1]
```diff
@@ -1426,14 +1426,47 @@ Constraint *PromiseAppendConstraint(Promise *pp, const char *lval, Rval rval, bo
...
+                    RlistAppendScalar(&synthetic_args, xstrdup(RvalScalarValue(old_cp->rval)));
+
+                    // append the old Rval (a function call) under the arguments of the new one
+                    RlistAppend(&synthetic_args, rval.item, RVAL_TYPE_FNCALL);
+
+                    Rval replacement = (Rval) { FnCallNew(xstrdup("and"), synthetic_args), RVAL_TYPE_FNCALL };
+                    Log(LOG_LEVEL_DEBUG, "PromiseAppendConstraint: MERGED %s rval %s", old_cp->lval, RvalToString(replacement));
+
+                    // overwrite the old Constraint rval with its replacement
+                    cp->rval = replacement;
```
Comment (kacfengine) [inline libpromises/policy.c]: Don't you need to ``RvalDestroy(cp->rval)`` this before assigning, like you do in the other case branch?

**Comment by kacfengine** [comment]
Added ``RvalDestroy`` and running through a private branch in order to test both fixes together, this one and https://github.com/cfengine/masterfiles/pull/503.

**Comment by kacfengine** [comment]
Both changes green, so merged.

**Comment by kacfengine** [comment]
Cherry-picked to 3.7.x as well.

**Comment by tzz** [comment]
Thanks for catching that memory leak and reviewing the whole thing.

**Pull request merged**